cloudevents/pkg/reconciler/events/cloudevent/cloudevent.go

```diff
@@ -0,0 +1,162 @@
+	PipelineRun *v1beta1.PipelineRun `json:"pipelineRun,omitempty"`
+}
+
+// newTektonCloudEventData returns a new instance of TektonCloudEventData
+func newTektonCloudEventData(runObject objectWithCondition) TektonCloudEventData {
+	tektonCloudEventData := TektonCloudEventData{}
+	switch v := runObject.(type) {
+	case *v1beta1.TaskRun:
+		tektonCloudEventData.TaskRun = v
+	case *v1beta1.PipelineRun:
+		tektonCloudEventData.PipelineRun = v
+	}
+	return tektonCloudEventData
+}
+
+func getEventType(runObject objectWithCondition) (*cdeevents.CDEventType, error) {
+	c := runObject.GetStatusCondition().GetCondition(apis.ConditionSucceeded)
+	if c == nil {
+		return nil, fmt.Errorf("no condition for ConditionSucceeded in %T", runObject)
+	}
+	var eventType cdeevents.CDEventType
+	switch {
+	// TODO : Update TaskRun/PipelineRun events from finished to success/fa
+	case c.IsUnknown():
+		switch runObject.(type) {
+		case *v1beta1.TaskRun:
+			switch c.Reason {
+			case v1beta1.TaskRunReasonStarted.String():
+				eventType = cdeevents.TaskRunStartedEventV1
+				//case v1beta1.TaskRunReasonRunning.String():
+				//	eventType = TaskRunRunningEventV1
+				//default:
+				//	eventType = TaskRunUnknownEventV1
+			}
+		case *v1beta1.PipelineRun:
+			switch c.Reason {
+			case v1beta1.PipelineRunReasonStarted.String():
+				eventType = cdeevents.PipelineRunStartedEventV1
+				//case v1beta1.PipelineRunReasonRunning.String():
+				//	eventType = PipelineRunRunningEventV1
+				//default:
+				//	eventType = PipelineRunUnknownEventV1
+			}
+		}
+	case c.IsFalse():
+		switch runObject.(type) {
+		case *v1beta1.TaskRun:
+			eventType = cdeevents.TaskRunFinishedEventV1 //TaskRunFailedEventV1
+		case *v1beta1.PipelineRun:
+			eventType = cdeevents.PipelineRunFinishedEventV1 //PipelineRunFailedEventV1
+		}
+	case c.IsTrue():
+		switch runObject.(type) {
+		case *v1beta1.TaskRun:
+			eventType = cdeevents.TaskRunFinishedEventV1 //TaskRunSuccessfulEventV1
+		case *v1beta1.PipelineRun:
+			eventType = cdeevents.PipelineRunFinishedEventV1 //PipelineRunSuccessfulEventV1
+		}
+	default:
+		return nil, fmt.Errorf("unknown condition for in %T.Status %s", runObject, c.Status)
+	}
+	return &eventType, nil
+}
+
+// eventForObjectWithCondition creates a new event based for a objectWithCondition,
+// or return an error if not possible.
+func eventForObjectWithCondition(runObject objectWithCondition) (*cloudevents.Event, error) {
+	event := cloudevents.NewEvent()
+	event.SetID(uuid.New().String())
+	event.SetSubject(runObject.GetObjectMeta().GetName())
+	// TODO: SelfLink is deprecated https://github.com/tektoncd/pipeline/issues/2676
+	source := runObject.GetObjectMeta().GetSelfLink()
+	if source == "" {
+		gvk := runObject.GetObjectKind().GroupVersionKind()
+		source = fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s/%s",
+			gvk.Group,
+			gvk.Version,
+			runObject.GetObjectMeta().GetNamespace(),
+			gvk.Kind,
+			runObject.GetObjectMeta().GetName())
```

Comment:
Uhm, this does not seem to work correctly, I wonder if it's broken in pipeline too?

Reply:
I get `/apis///namespaces/default//guarded-pr-tmxzd`, so a few fields are missing